// ffaivideo/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Logger = exports.SubMaker = exports.defaultLLMConfig = exports.VoiceConfig = exports.VideoConcatMode = exports.VideoAspect = exports.generateVideo = exports.getAxiosConfig = exports.buildApiUrl = exports.httpGet = exports.getSize = exports.setSize = exports.setWH = exports.getWH = void 0;
const generator_1 = require("./generator");
Object.defineProperty(exports, "generateVideo", { enumerable: true, get: function () { return generator_1.generateVideo; } });
const constant_1 = require("./config/constant");
Object.defineProperty(exports, "VideoAspect", { enumerable: true, get: function () { return constant_1.VideoAspect; } });
Object.defineProperty(exports, "VideoConcatMode", { enumerable: true, get: function () { return constant_1.VideoConcatMode; } });
const llm_config_1 = require("./config/llm-config");
Object.defineProperty(exports, "defaultLLMConfig", { enumerable: true, get: function () { return llm_config_1.defaultLLMConfig; } });
const voice_config_1 = require("./config/voice-config");
Object.defineProperty(exports, "VoiceConfig", { enumerable: true, get: function () { return voice_config_1.VoiceConfig; } });
const video_aspect_1 = require("./config/video-aspect");
Object.defineProperty(exports, "getWH", { enumerable: true, get: function () { return video_aspect_1.getWH; } });
Object.defineProperty(exports, "setWH", { enumerable: true, get: function () { return video_aspect_1.setWH; } });
Object.defineProperty(exports, "setSize", { enumerable: true, get: function () { return video_aspect_1.setSize; } });
Object.defineProperty(exports, "getSize", { enumerable: true, get: function () { return video_aspect_1.getSize; } });
const http_1 = require("./utils/http");
Object.defineProperty(exports, "httpGet", { enumerable: true, get: function () { return http_1.httpGet; } });
Object.defineProperty(exports, "buildApiUrl", { enumerable: true, get: function () { return http_1.buildApiUrl; } });
Object.defineProperty(exports, "getAxiosConfig", { enumerable: true, get: function () { return http_1.getAxiosConfig; } });
const log_1 = require("./utils/log");
Object.defineProperty(exports, "Logger", { enumerable: true, get: function () { return log_1.Logger; } });
const sub_maker_1 = require("./sub-maker");
Object.defineProperty(exports, "SubMaker", { enumerable: true, get: function () { return sub_maker_1.SubMaker; } });
