// ffaivideo/dist/config/constant.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PUNCTUATIONS = exports.successFun = exports.progressFun = exports.VideoConcatMode = exports.VideoAspect = exports.TINY_PUNCTUATIONS = void 0;
const PUNCTUATIONS = [
    '?',
    ',',
    '.',
    ';',
    ':',
    '!',
    '…',
    '？',
    '，',
    '。',
    '、',
    '；',
    '：',
    '！',
    '...',
];
exports.PUNCTUATIONS = PUNCTUATIONS;
const TINY_PUNCTUATIONS = PUNCTUATIONS.filter(punctuation => !['、', ':', '：'].includes(punctuation));
exports.TINY_PUNCTUATIONS = TINY_PUNCTUATIONS;
var VideoAspect;
(function (VideoAspect) {
    VideoAspect["Landscape"] = "16:9";
    VideoAspect["Portrait"] = "9:16";
    VideoAspect["Square"] = "1:1";
})(VideoAspect || (exports.VideoAspect = VideoAspect = {}));
var VideoConcatMode;
(function (VideoConcatMode) {
    VideoConcatMode["Sequential"] = "sequential";
    VideoConcatMode["Random"] = "random";
})(VideoConcatMode || (exports.VideoConcatMode = VideoConcatMode = {}));
const progressFun = (p) => { };
exports.progressFun = progressFun;
const successFun = (r) => { };
exports.successFun = successFun;
