// ffaivideo/dist/config/llm-config.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.defaultLLMConfig = void 0;
const defaultLLMConfig = {
    moonshot: {
        apiKey: '',
        modelName: 'moonshot-v1-8k',
        baseUrl: 'https://api.moonshot.cn/v1',
    },
    openai: {
        apiKey: '',
        modelName: 'gpt-4-turbo-preview',
        baseUrl: 'https://api.openai.com/v1',
    },
    azure: {
        apiKey: '',
        modelName: 'gpt-35-turbo',
        baseUrl: '***',
        apiVersion: '2024-02-15-preview',
    },
    gemini: {
        apiKey: '',
        modelName: '',
        baseUrl: '***',
    },
    g4f: {
        apiKey: 'xxx',
        modelName: 'gpt-3.5-turbo-16k-0613',
        baseUrl: '***',
    },
    gpt4js: {
        apiKey: 'xxx',
        provider: "Nextway",
        modelName: 'gpt-4o-free',
        baseUrl: '***',
    },
    customAI: {
        apiKey: '',
        modelName: '',
        baseUrl: '',
    },
    defalut: {
        apiKey: '',
        modelName: '',
        baseUrl: '***',
    },
};
exports.defaultLLMConfig = defaultLLMConfig;
